### DISCUSSION TITLE
Cache busting

### DISCUSSION THREAD

**Discussion by sluc69** [joomla/joomla-cms#34372] [Q&A] [answered]
Is there future work on versioning of file-types like CSS, JS images? 
Whenever a file content change the version attached to the file should change so the client browser cache is updated.
https://www.keycdn.com/support/what-is-cache-busting#:~:text=Cache%20busting%20solves%20the%20browser,server%20for%20the%20new%20file.

**Comment by richard67** [comment]
@dgrammatiko I guess that's your topic 😄

**Reply by dgrammatiko** [reply]
This is already implemented in Joomla 4 using the 3rd option of the linked document `Query strings (e.g. style.css?ver=2)`.

In fact, this was existing for way too long in Joomla what was changed in version 4 was the way the query string was created:
- in J3 the query string was changing for each new version
- in J4 the query string reflects the hash of the file

FWIW the hash is not evaluated at run time but is precalculated at build time, through Joomla's build tools, and saved with all the other related information in the relative joomla.asset.json: https://docs.joomla.org/Special:MyLanguage/J4.x:Web_Assets
The actual thing looks like this: 
![Screenshot 2021-06-03 at 20 25 20](https://user-images.githubusercontent.com/3889375/120693935-e7a80f00-c4a9-11eb-88ff-a654366f5bcd.png)

Also, another thing that Joomla 4 is doing is that it provides pre-compressed static assets (.gz) so the server just sends the file without the need to compress it at run time.

It would be a great feature if the extension directory would set some minimum requirements to what a J4 ready extension for the benefit of all the users (but I won't hold my breath on this)


PS: This technic atm is only affecting css/js files.
There is a PR for fonts and images in css: https://github.com/joomla/joomla-cms/pull/32574 and at some point, there will be a solution for images (assuming PRs will be accepted/merged)

**Reply by sluc69** [reply]
I can see the query string on CSS files and the NONCE on inline-scripts but the query string on CSS doesn't change when the file is modified. The query string is the same for all CSS files. Do I have to jump through hoops to get that changed?? This is from a USER perspective! 

![Screenshot 2021-06-04 145359](https://user-images.githubusercontent.com/82952282/120758709-bdeff600-c544-11eb-8cb1-e233771d990e.jpg)


At the link in my question it also reads:
_Query strings on the other hand, have been known to cause caching issues. Some proxies or CDNs aren't even able to cache files that contain query strings and it is recommended not to use them._

**Reply by dgrammatiko** [reply]
> but the query string on CSS doesn't change when the file is modified

A couple of things here:

- You cannot **edit** static assets files that you don't own, eg you're not the dev of the extension or Joomla unless you override them in the template because they will be overwritten in the next update
- The system is not doing the content hashing at run time as I wrote above but rather at build time (per release)
- - the system excludes the templates atm, the plan is to bring only the new ones (the ones that support child templates)
- The system I described is opt-in, 3rd PD are not obliged to implement it although I would say that for the end-users benefit that should be an acceptance rule in the extensions directory
- Creating and editing an override, atm, will not reflect the changes, there's a bigger task to finalise the child templates with all shenanigans (eg rewrite the com_templates)

If you need to bust manually the browser's caches at your own pace (editing files manually) there's a simple module for that: https://invalidate-cache.dgrammatiko.dev

**Reply by dgrammatiko** [reply]
> Query strings on the other hand, have been known to cause caching issues. Some proxies or CDNs aren't even able to cache files that contain query strings and it is recommended not to use them.

That's fine for Joomla as none of the local assets are either on a CDN (they live at your server) and probably the site isn't behind a proxy (although even if it is you as the owner have all the rights to choose how that proxy will treat this url). In short there's no probleme with this strategy

**Comment by richard67** [comment]
@sluc69 Does @dgrammatiko 's reply above answer your question? If so, please mark it as answer by using the check mark icon at the top right corner of that post, or let us know then we can do that. Thanks in advance.

**Reply by sluc69** [reply]
I  understand I cannot edit static assets files that you don't own, and I use overrides to edit them.

Perhaps you have seen the <casionopeia template>/user.css... and the <casionopeia template>/user.js
You are saying I should write anything in these files??? Then what is the purpose of these files and how can I add styles and javascript to the site???

When I upgraded to RC1 these files were NOT overridden. The site worked as before the upgrade. I just had to redo the CSP.

Do you understand the problem I present about USERS adding style and jscripts (avoiding as much as possible inline).

But thank you for pointing me to the add-on. It solves the problem for me now. 
According the add-on website, this function is already present in J4. Just needs activation.....

**Reply by dgrammatiko** [reply]
> Do you understand the problem I present about USERS adding style and jscripts (avoiding as much as possible inline).

I do! The only efficient way to solve this is to create the hash at the time of the file saving and we're back to my comment that the component that handles the templates is not ready/updated yet: https://github.com/joomla/joomla-cms/pull/32896

> But thank you for pointing me to the add-on. It solves the problem for me now.

Glad to hear that. I wrote that module knowing that would be helpful...

**Comment by richard67** [comment]
Pity, we can only select comments on the first level as answer, so we can't mark Dimitris' comments as the answer.

**Comment by sluc69** [comment, marked as answer]
How is cache-busting implemented now in J4??
The 'invalidate cache module' is not working anymore since J4.0x
Is there any way to invalidate images? like uploading a changed image with the same name

**Reply by dgrammatiko** [reply]
> The 'invalidate cache module' is not working anymore since J4.0x

I will check that, thanks for reporting. Just checked and the 4.x version works fine for me

> Is there any way to invalidate images? like uploading a changed image with the same name

Well, there's again a plugin that solves the image cache invalidation: https://responsive-images.dgrammatiko.dev

**Reply by sluc69** [reply]
what version of invalidate nodule you use? On 2 sites (working before) I get no response (green alert) anymore, even uninstalled reinstalled. On one site I get "something blew up" red alert message

**Reply by dgrammatiko** [reply]
The 3.x will not work on Joomla 4 and vice versa. My test was on https://invalidate-cache.dgrammatiko.dev/dist/j4/mod_invalidatecache_0.0.8.zip

**Reply by sluc69** [reply]
Ok I found the reason; Using sidekick browser is causing the problem! other browsers work correctly.
Will try the Responsive image plugin. 
Thanks
